src/Rules/Names/UsedNamesRule.php

```diff
@@ -0,0 +1,144 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Rules\Names;
+
+use PhpParser\Node;
+use PhpParser\Node\Stmt\ClassLike;
+use PhpParser\Node\Stmt\Enum_;
+use PhpParser\Node\Stmt\GroupUse;
+use PhpParser\Node\Stmt\Interface_;
+use PhpParser\Node\Stmt\Namespace_;
+use PhpParser\Node\Stmt\Trait_;
+use PhpParser\Node\Stmt\Use_;
+use PhpParser\Node\Stmt\UseUse;
+use PHPStan\Analyser\Scope;
+use PHPStan\Node\FileNode;
+use PHPStan\Rules\Rule;
+use PHPStan\Rules\RuleError;
+use PHPStan\Rules\RuleErrorBuilder;
+use function array_merge;
+use function in_array;
+use function sprintf;
+use function strtolower;
+
+/**
+ * @implements Rule<FileNode>
+ */
+final class UsedNamesRule implements Rule
+{
+
+	public function getNodeType(): string
+	{
+		return FileNode::class;
+	}
+
+	/**
+	 * @param FileNode $node
+	 */
+	public function processNode(Node $node, Scope $scope): array
+	{
+		$usedNames = [];
+		$errors = [];
+		foreach ($node->getNodes() as $oneNode) {
+			if ($oneNode instanceof Namespace_) {
+				$namespaceName = $oneNode->name !== null ? $oneNode->name->toString() : '';
+				foreach ($oneNode->stmts as $stmt) {
+					$errors[] = $this->findErrorsForNode($stmt, $namespaceName, $usedNames);
+				}
+			}
+
+			$errors[] = $this->findErrorsForNode($oneNode, '', $usedNames);
+		}
+
+		return array_merge(...$errors);
```

Comment:
What is this supposed to do? Can't we just `return $errors;`?

Reply:
No, $errors is array<RuleError[]>. I wanted to avoid to call array_merge in foreach. See https://www.exakat.io/en/speeding-up-array_merge/

Reply:
In that case please just iterate over `$this->findErrorsForNode($oneNode, '', $usedNames)` each time and do `$errors[] = $error;`

Reply:
ok no problem

Reply:
done